### PR TITLE
Enable adam foreach in full train step tracing

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97897
* #97853

Main changes:

1. Registered several foreach ops to both meta and DTensor
2. Skip redundant getitem node when expanding foreach ops with DTensor

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -1206,14 +1232,75 @@ def _check_foreach_binop_tensor_lists(self, other):
...
+    )
+    check(
+        len(self) == len(tensor1) and len(self) == len(tensor2),
+        lambda: "All input tensor lists must have the same length",
+    )
+
+
+@register_meta([aten._foreach_pow.ScalarAndTensor])
+def meta__foreach_pow_scalar_and_tensor(self, exponent):
+    check(isinstance(exponent, List), lambda: f"exponent must be a tensor list but got {type(exponent)}")
```
Comment (bdhirsh) [inline torch/_meta_registrations.py:1302]: these all look ok to me.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -1206,14 +1232,75 @@ def _check_foreach_binop_tensor_lists(self, other):
...
+
+@register_meta(
+    [
+        aten._foreach_addcdiv_.Scalar,
+        aten._foreach_addcmul_.Scalar,
+    ]
+)
+def meta__foreach_addcop__scalar(self, tensor1, tensor2, scalar=1):
+    assert all([isinstance(l, List) for l in [self, tensor1, tensor2]]), (
+        "All arguments of _foreach_addcmul_ must be List[Tensor], but got "
```
Comment (bdhirsh) [inline torch/_meta_registrations.py:1287]: Nit: don't hardcode `addcmul_` into the error string if you're re-using this meta kernel for multiple ops.

If we need the hardcoded string (to satisfy tests), you'll have to register the function separately.

File: torch/distributed/_spmd/experimental_ops.py [hunk 3]
```diff
@@ -17,17 +17,45 @@
...
+        aten._foreach_sqrt.default,
+    ]
+)
+def _prop__foreach_unaop(op_schema: OpSchema) -> OutputSharding:
+    self = op_schema.args_schema[0]
     assert isinstance(self, list) and all(
         [isinstance(s, DTensorSpec) for s in self]
     )
+    # FIXME(@mrshenli): for sqrt, this is only mathematically correct for
+    # Replicate and Shard tensor.
```
Comment (wanchaol) [inline torch/distributed/_spmd/experimental_ops.py:33]: oh do you mean it needs to guard on partial?

**Review by wanchaol — APPROVED** [review]

File: torch/distributed/_spmd/experimental_ops.py (hunk 3 above)
Comment (mrshenli) [reply torch/distributed/_spmd/experimental_ops.py:33]: not sure yet, we might need to return arg schema suggestions with either Replicate or Shard to make this sound, but it doesn't seem the optimal decision on Replicate vs Shard can be make locally here?

**Review by fegin — APPROVED** [review]
LGTM

**Pull request closed without merging**